File: src/main/java/com/mk7a/soulbind/listeners/CommandBlockerListener.java
package com.mk7a.soulbind.listeners;

import com.mk7a.soulbind.main.ItemSoulBindPlugin;
import com.mk7a.soulbind.main.PluginConfiguration;
import com.mk7a.soulbind.util.BindUtil;
import com.mk7a.soulbind.util.Util;
import org.bukkit.Material;
import org.bukkit.entity.Player;
import org.bukkit.event.EventHandler;
import org.bukkit.event.EventPriority;
import org.bukkit.event.Listener;
import org.bukkit.event.player.PlayerCommandPreprocessEvent;
import org.bukkit.inventory.ItemStack;
import org.bukkit.inventory.PlayerInventory;

public class CommandBlockerListener implements Listener {

    private final ItemSoulBindPlugin plugin;
    private final PluginConfiguration config;

    public CommandBlockerListener(ItemSoulBindPlugin plugin) {
        this.plugin = plugin;
        config = ItemSoulBindPlugin.getPluginConfig();
    }

    public void register() {
        plugin.getServer().getPluginManager().registerEvents(this, plugin);
    }

    @EventHandler(priority = EventPriority.HIGH)
    public void preCommand(PlayerCommandPreprocessEvent event) {

        Player player = event.getPlayer();
        PlayerInventory inventory = player.getInventory();

        if (isCommandBlocked(event.getMessage())) {

            ItemStack mainHandItem = inventory.getItemInMainHand();

            if (!mainHandItem.getType().equals(Material.AIR)) {


                if (BindUtil.hasBind(mainHandItem)) {

                    event.setCancelled(true);
                    Util.sendMessage(player, config.cmdBlocked);
                }
            }

        }
    }

    private boolean isCommandBlocked(String fullCommand) {

        String command = fullCommand.substring(1);

        String mainCommand = command.split(" ")[0];
        boolean simpleCheck = config.blockedCommands.contains(mainCommand);

        boolean advancedCheck = false;

        for (String blocked : config.blockedCommands) {
            if (command.toLowerCase().startsWith(blocked.toLowerCase())) {
                advancedCheck = true;
                break;
            }
        }

        return simpleCheck || advancedCheck;
    }
}

File: src/main/kotlin/com/mk7a/soulbind/listener/CommandBlockerListener.kt
package com.mk7a.soulbind.listener

import com.mk7a.soulbind.ItemSoulBindPlugin
import com.mk7a.soulbind.bind.isSoulBound
import org.bukkit.event.EventHandler
import org.bukkit.event.EventPriority
import org.bukkit.event.Listener
import org.bukkit.event.player.PlayerCommandPreprocessEvent

/** Blocks configured commands (e.g. auction listings) while a soul bound item is held. */
class CommandBlockerListener(private val plugin: ItemSoulBindPlugin) : Listener {

    private val settings get() = plugin.settings

    @EventHandler(priority = EventPriority.HIGH)
    fun onCommand(event: PlayerCommandPreprocessEvent) {
        val command = event.message.removePrefix("/")
        if (settings.blockedCommands.none { command.startsWith(it, ignoreCase = true) }) return

        val player = event.player
        if (player.inventory.itemInMainHand.isSoulBound) {
            event.isCancelled = true
            player.sendMessage(settings.messages.commandBlocked)
        }
    }
}
